File: 907-koko-eating-bananas/koko-eating-bananas.ts
function minEatingSpeed(piles: number[], h: number): number {
    let l = 1;
    let r = Math.max(...piles);
    let ans = Infinity;

    function getHoursGivenRate(r: number): number {
        let hours = 0;
        for (let i = 0; i < piles.length; i++) {
            hours += Math.ceil(piles[i] / r);
        }
        return hours;
    }

    while (l <= r) {
        let m = Math.floor((l + r) / 2);
        let hours = getHoursGivenRate(m);
        if (hours <= h) {
            ans = Math.min(ans, m);
            r = m - 1;
        }
        else l = m + 1;
    }
    return ans;

};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int r = 0;
        int ans = -1;
        for(int pile : piles)r = Math.max(r , pile);
        int l = 1;
        while(l <= r) {
            int m = l + (r - l) / 2;
            int hr = findHoursGivenRate(piles , m);
            if(hr > h) l = m + 1;
            else {
                ans = m;
                r = m - 1;
            }
        }
        return ans;
    }
    private int findHoursGivenRate(int[] piles , int k) {
        int h = 0;
        for(int i = 0; i < piles.length; i ++) {
            h += Math.ceil((double) piles[i] / k);
        }
        return h;
    }
}